src/cmds/pbsnodes.c

```diff
@@ -244,110 +237,63 @@ encode_to_json(struct batch_status *bstat) {
+								return 1;
 					} else if (str[0] != '0') {
-						node = add_json_node(JSON_VALUE, JSON_STRING, next->resource, str);
-						if (node == NULL)
+						if (add_json_node(JSON_VALUE, JSON_STRING, next->resource, str) == NULL)
 							return 1;
 					}
 				} else {
 					if (str[0] != '0') {
-						node = add_json_node(JSON_VALUE, JSON_STRING, next->resource, str);
-						if (node == NULL)
+						if (add_json_node(JSON_VALUE, JSON_STRING, next->resource, str) == NULL)
 							return 1;
 					}
 				}
-				if (next->next) {
-					if (strcmp(next->next->name, "resources_assigned") == 0)
-						next = next->next;
-					else {
-						/* In case of an empty object */
-						if (node->node_type == JSON_OBJECT) {
-							node = add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL);
-							if (node == NULL)
-								return 1;
-						} else {
-							node->node_type = JSON_OBJECT_END;
-							pattr = next;
-							next = NULL;
-							break;
-						}
-					}
+				if (next->next == NULL || strcmp(next->next->name, "resources_assigned")) {
+					/* Nothing left in resources_assigned, close object */
+					if (add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL) == NULL)
+						return 1;
+					pattr = next;
+					next = NULL;
 				} else {
-					/* In case of an empty object adding an extra node to end the object. */
-					if (node->node_type == JSON_OBJECT) {
-						node = add_json_node(JSON_OBJECT_END, JSON_NULL, NULL, NULL);
-						if (node == NULL)
-							return 1;
-					} else {
-						node->node_type = JSON_OBJECT_END;
-						pattr = next;
-						next = NULL;
-						break;
-					}
+					next = next->next;
 				}
 			}
-		} else {
-			node = add_json_node(JSON_VALUE, JSON_NULL, pattr->name, NULL);
-			if (node == NULL)
+		} else if (strcmp(pattr->name, "jobs") == 0) {
+			if (add_json_node(JSON_ARRAY, JSON_NULL, pattr->name, NULL) == NULL)
 				return 1;
-			str = pattr->value;
-			value = strtod(pattr->value, &pc);
-			while (pc) {
-				if (isspace(*pc))
-					pc++;
-				else
-					break;
+			pc = pattr->value;
+			pc = pc1 = str;
+			while (*pc1) {
+				if (*pc1 != ' ')
+					*pc++ = *(pc1);
+				pc1++;
 			}
-			if (strcmp(pc, "") == 0) {
-				ivalue = (long int)value;
-				if (value == ivalue) {
-					node->value_type = JSON_INT;
-					node->value.inumber = ivalue;
-				} else {
-					node->value_type = JSON_FLOAT;
-					node->value.fnumber = value;
-				}
-			} else if (strcmp(pattr->name, "jobs")== 0) {
-				node->node_type = JSON_ARRAY;
-				pc = pc1 = str;
-				while (*pc1) {
-					if (*pc1 != ' ')
-						*pc++ = *(pc1);
-					pc1++;
-				}
-				*pc = '\0';
-				pc = strtok(str, ",");
-				while (pc) {
-					pc1 = strchr(pc, '/');
-					if(pc1)
-						*pc1 = '\0';
-					if (strcmp(pc, prev_jobid) != 0) {
-						node = add_json_node(JSON_VALUE, JSON_STRING, NULL, pc);
-						if (node == NULL)
-							return 1;
-					}
-					prev_jobid = pc;
-					pc = strtok(NULL, ",");
+			*pc = '\0';
+			pc = strtok(str, ",");
+			while (pc) {
```

Comment:
You're iterating with strtok() here, turn this into a for() loop.

Reply:
Changed